### PR TITLE
Implement BearerAuthentication class.

### PR DESCRIPTION
**Pull request by awais786**
WIP: Add BearerAuthentication.
Add test cases.

### CODE REVIEW

**Comment by awais786** [comment]
@rlucioni Is there any thing else I have to add in this PR ?

File: credentials/apps/api/authentication.py [hunk 1]
```diff
@@ -63,3 +68,61 @@ def authenticate_credentials(self, payload):
...
+
+class BearerAuthentication(BaseAuthentication):
+    """
+    Simple token based authentication.
+    Clients should authenticate by passing the token key in the "Authorization"
+    HTTP header, prepended with the string "Bearer ".  For example:
+        Authorization: Bearer 401f7ac837da42b97f613d789819ff93537bee6a
+    """
+
+    def authenticate(self, request):
```
Comment (ahsan-ul-haq) [inline credentials/apps/api/authentication.py]: Please add Doc string

File: acceptance_tests/mixins.py [hunk 2]
```diff
@@ -21,3 +22,13 @@ def login_with_lms(self):
 
         self.lms_login_page.browser.get(self.lms_login_page.url())  # pylint: disable=not-callable
         self.lms_login_page.login(email, password)
+
+
+class CredentialsApiMixin(object):
+    """ Mixin used for login on credentials."""
+    def setUp(self):
```
Comment (ahsan-ul-haq) [inline acceptance_tests/mixins.py]: What is reason of having this setup?

File: credentials/apps/api/tests/test_authentication.py [hunk 3]
```diff
@@ -1,22 +1,29 @@
...
 
 import ddt
+from django.conf import settings
 from django.contrib.auth.models import Group
-from django.test import TestCase
+from django.test import override_settings, RequestFactory, TestCase
 import mock
 from rest_framework.exceptions import AuthenticationFailed
 from rest_framework.test import APIRequestFactory
+import httpretty
```
Comment (ahsan-ul-haq) [inline credentials/apps/api/tests/test_authentication.py]: Alphabetize

File: credentials/apps/api/tests/test_authentication.py [hunk 4]
```diff
@@ -1,22 +1,29 @@
...
 
-from credentials.apps.api.authentication import JwtAuthentication, pipeline_set_user_roles
+from credentials.apps.api.authentication import JwtAuthentication, pipeline_set_user_roles, BearerAuthentication
 from credentials.apps.api.jwt_decode_handler import api_settings as drf_jwt_settings
 from credentials.apps.api.tests.mixins import JwtMixin
 from credentials.apps.core.constants import Role
+from credentials.apps.core.tests import factories
 from credentials.apps.core.tests.factories import UserFactory
```
Comment (ahsan-ul-haq) [inline credentials/apps/api/tests/test_authentication.py]: If you are importing factories then remove this import

File: credentials/apps/api/tests/test_authentication.py [hunk 5]
```diff
@@ -146,3 +153,102 @@ def test_no_user(self):
...
+        self.factory = RequestFactory()
+
+    def _create_request(self, token=AccessTokenMixin.DEFAULT_TOKEN, token_name='Bearer'):
+        """Create request with authorization header. """
+        auth_header = '{} {}'.format(token_name, token)
+        request = self.factory.get('/', HTTP_AUTHORIZATION=auth_header)
+        return request
+
+    def test_authenticate_header(self):
+        """ The method should return the string Bearer. """
```
Comment (ahsan-ul-haq) [inline credentials/apps/api/tests/test_authentication.py]: No need to add spaces at start and end of docstring.

File: acceptance_tests/mixins.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 import logging
+from edx_rest_api_client.client import EdxRestApiClient
```
Comment (rlucioni) [inline acceptance_tests/mixins.py]: This is a third-party package, not part of the standard library. Please put it in the appropriate import block.

File: credentials/apps/api/authentication.py [hunk 7]
```diff
@@ -3,9 +3,14 @@
 """
 
 import logging
+import requests
```
Comment (rlucioni) [inline credentials/apps/api/authentication.py]: This is a third-party package, not part of the standard library. Please put it in the appropriate import block.

File: credentials/apps/api/authentication.py [hunk 8]
```diff
@@ -3,9 +3,14 @@
...
 
+from django.conf import settings
 from django.contrib.auth.models import Group
+
```
Comment (rlucioni) [inline credentials/apps/api/authentication.py]: Delete this line. Django and DRF are both third-party packages.

File: credentials/apps/api/authentication.py [hunk 9]
```diff
@@ -63,3 +68,61 @@ def authenticate_credentials(self, payload):
             user.groups.remove(admin_group)
 
         return user
+
+
+class BearerAuthentication(BaseAuthentication):
+    """
+    Simple token based authentication.
```
Comment (rlucioni) [inline credentials/apps/api/authentication.py]: Move this up a line so it's right next to the opening block quote.

File: credentials/apps/api/authentication.py [hunk 10]
```diff
@@ -63,3 +68,61 @@ def authenticate_credentials(self, payload):
...
+    Clients should authenticate by passing the token key in the "Authorization"
+    HTTP header, prepended with the string "Bearer ".  For example:
+        Authorization: Bearer 401f7ac837da42b97f613d789819ff93537bee6a
```
Comment (rlucioni) [inline credentials/apps/api/authentication.py]: Put an empty line above this line to space out the docstring, making it easier to read.

File: credentials/apps/api/authentication.py (hunk 1 above)
Comment (rlucioni) [inline credentials/apps/api/authentication.py]: @ahsan-ul-haq I think it's okay not to have one here. This method is a standard part of any DRF authentication class.

File: credentials/apps/api/tests/test_authentication.py [hunk 11]
```diff
@@ -1,22 +1,28 @@
...
 from django.contrib.auth.models import Group
-from django.test import TestCase
+from django.test import override_settings, RequestFactory, TestCase
+import httpretty
 import mock
 from rest_framework.exceptions import AuthenticationFailed
 from rest_framework.test import APIRequestFactory
 
-from credentials.apps.api.authentication import JwtAuthentication, pipeline_set_user_roles
+from credentials.apps.api.authentication import JwtAuthentication, pipeline_set_user_roles, BearerAuthentication
```
Comment (rlucioni) [inline credentials/apps/api/tests/test_authentication.py]: Nitpick: 
```python
from credentials.apps.api.authentication import BearerAuthentication, JwtAuthentication, pipeline_set_user_roles
```

File: credentials/apps/api/tests/test_authentication.py [hunk 12]
```diff
@@ -1,22 +1,28 @@
...
 from credentials.apps.api.jwt_decode_handler import api_settings as drf_jwt_settings
 from credentials.apps.api.tests.mixins import JwtMixin
 from credentials.apps.core.constants import Role
 from credentials.apps.core.tests.factories import UserFactory
 
 
+OAUTH2_PROVIDER_URL = 'https://example.com/oauth2'
```
Comment (rlucioni) [inline credentials/apps/api/tests/test_authentication.py]: Why not just get this from (test) settings?

File: credentials/apps/api/tests/test_authentication.py (hunk 5 above)
Comment (rlucioni) [inline credentials/apps/api/tests/test_authentication.py]: Seconding this. It's a nitpick, but I'd remove them.

File: credentials/apps/api/tests/test_authentication.py [hunk 13]
```diff
@@ -146,3 +152,101 @@ def test_no_user(self):
...
+class BearerAuthenticationTests(AccessTokenMixin, TestCase):
+    """
+    Test Simple token based authentication used with the browseable API.
+    """
+    def setUp(self):
+        super(BearerAuthenticationTests, self).setUp()
+        self.auth = BearerAuthentication()
+        self.factory = RequestFactory()
+
+    def _create_request(self, token=AccessTokenMixin.DEFAULT_TOKEN, token_name='Bearer'):
```
Comment (rlucioni) [inline credentials/apps/api/tests/test_authentication.py]: `token_name` should be `header`.

Comment (rlucioni) [inline credentials/apps/api/tests/test_authentication.py]: Whoops, ignore this. Sorry.

**Comment by rlucioni** [comment]
@awais786 just some minor cleanup comments. :+1: once they're addressed.

I think it's beyond the scope of this PR, but we may want to consider moving this authentication class into a separate package. It's broadly useful.

**Comment by awais786** [comment]
@rlucioni would you please explain how to move this into a separate package. Moving into some other location ?

**Comment by rlucioni** [comment]
@awais786 I'd move it to a separate repo and upload the package to PyPI. For example, see https://github.com/jpadilla/django-rest-framework-oauth. That package contains authentication classes like [`OAuth2Authentication`](https://github.com/jpadilla/django-rest-framework-oauth/blob/master/rest_framework_oauth/authentication.py#L122) which we use in edx-platform.

As I said, I don't think it's something we need to do now. However, we should keep it in mind since it would allow us to reduce the amount of code duplicated across IDAs.

**Comment by awais786** [comment]
@rlucioni I have updated the tests cases. As it is added DEFAULT_AUTHENTICATION_CLASSES so we need to consider some testing/security concerns here ?

**Comment by awais786** [comment]
I am adding new story to implement this as a separate package.

**Comment by rlucioni** [comment]
@awais786 what would you do differently? I think what you've done is fine. :+1:

**Comment by awais786** [comment]
@rlucioni I am planning to pull this story into this sprint. Please let me know if you have any concerns.

**Comment by rlucioni** [comment]
@awais786 I'm not sure which story you're referring to. If it's pulling this authentication class into a separate package, that was just done. See https://github.com/edx/edx-drf-extensions/blob/master/edx_rest_framework_extensions/authentication.py. You could modify credentials so that it uses that class instead of implementing its own.

**Comment by awais786** [comment]
@rlucioni I was talking about this [ECOM-3808](https://openedx.atlassian.net/browse/ECOM-3808)

**Pull request closed without merging**